File: core/java/src/main/java/hara/lang/protocol/IWorkRun.java
package hara.lang.protocol;

import hara.lang.declaration.HaraAvailability;
import hara.lang.declaration.HaraMethod;
import hara.lang.declaration.HaraProtocolBinding;

/** Live work run with asynchronous result, events, and cancellation. */
@HaraProtocolBinding(
    namespace = "std.protocol.iworkrun",
    name = "IWorkRun",
    parents = {"IWorkRef", "IClosed"},
    availability = HaraAvailability.CAPABILITY_GATED,
    capability = "native-runtime-protocols")
public interface IWorkRun extends IWorkRef, IClosed {
  @HaraMethod(value = "work-status", arity = 1)
  Object workStatus();

  @HaraMethod(value = "work-result", arity = 1)
  IPromise workResult();

  @HaraMethod(value = "work-events", arity = 2)
  IStream workEvents(Object options);

  @HaraMethod(value = "work-cancel", arity = 2)
  IPromise workCancel(Object reason);
}

File: core/rust/src/lang/protocol/iworkrun.rs
use super::{IClosed, IStream, IWorkRef};
use hara_protocol_macros::hara_protocol;

#[hara_protocol(
    namespace = "std.protocol.iworkrun",
    name = "IWorkRun",
    parents = ["IWorkRef", "IClosed"],
    availability = "capability-gated",
    capability = "native-runtime-protocols"
)]
pub trait IWorkRun: IWorkRef + IClosed {
    type Status;
    type Promise;
    type Options;
    type Stream: IStream;
    type Reason;

    #[hara_method(value = "work-status", arity = 1)]
    fn work_status(&self) -> Self::Status;
    #[hara_method(value = "work-result", arity = 1)]
    fn work_result(&self) -> Self::Promise;
    #[hara_method(value = "work-events", arity = 2)]
    fn work_events(&self, options: Self::Options) -> Self::Stream;
    #[hara_method(value = "work-cancel", arity = 2)]
    fn work_cancel(&self, reason: Self::Reason) -> Self::Promise;
}
